### PR TITLE
#3260 - CAS Integration 3A - Create new Supplier and Site - Retry on Error and Scheduler Time

### PR DESCRIPTION
**Pull request by andrewsignori-aot**
- Changed the to allow throwing an error when some error is present in the process summary error attending the below AC.
```
If some error is present throw in the scheduler to allow it to be retried.
```
- Added the below E2E to test the above change.
  - Should throw an error for the first student and process the second one when the CAS API call failed for the first student but worked for the second one.
- Adjusted the scheduler to be executed during weekdays only attending the below AC (`0 20 * * 1-5`).
```
Schedule these to go once a day at noon during weekdays for now.
```

### Rollback
image

### Fix
- `COUNTRY_CANADA` was defined as "canada" when it is defined as "Canada" is all form.io definitions leading to an incorrect logic.
- SFTP issue while providing the file content as a string instead of a Buffer as identified by @dheepak-aot.

### CODE REVIEW

File: sources/packages/backend/apps/queue-consumers/src/processors/schedulers/cas-integration/cas-supplier-integration.scheduler.ts [hunk 1]
```diff
@@ -63,19 +62,16 @@ export class CASSupplierIntegrationScheduler extends BaseScheduler<void> {
           `Records updated: ${suppliersUpdated}.`,
         ],
         processSummary,
+        { throwOnError: true },
```
Comment (sh16011993) [inline sources/packages/backend/apps/queue-consumers/src/processors/schedulers/cas-integration/cas-supplier-integration.scheduler.ts:65]: 👍

File: sources/packages/backend/apps/queue-consumers/src/processors/schedulers/cas-integration/cas-supplier-integration.scheduler.ts [hunk 2]
```diff
@@ -63,19 +62,16 @@ export class CASSupplierIntegrationScheduler extends BaseScheduler<void> {
...
-        if (error.name === CAS_AUTH_ERROR) {
-          // Throw an error to force the queue to retry.
-          throw new Error(
-            `Unable to request data to CAS due to an authentication error.`,
-          );
-        }
-        const errorMessage = "Unexpected error while executing the job.";
-        processSummary.error(errorMessage, error);
-        return [errorMessage];
+        throw new Error(error.message);
```
Comment (sh16011993) [inline sources/packages/backend/apps/queue-consumers/src/processors/schedulers/cas-integration/cas-supplier-integration.scheduler.ts:70]: We won't throw the cause here? `cause: error`

Comment (andrewsignori-aot) [reply sources/packages/backend/apps/queue-consumers/src/processors/schedulers/cas-integration/cas-supplier-integration.scheduler.ts:70]: For custom-named errors, we assume the message will be friendly.

**Review by sh16011993 — APPROVED** [review]
Great work @andrewsignori-aot  👍  Thank You for explaining the mocks related to the e2e test. Very helpful.

File: sources/packages/backend/apps/db-migrations/src/sql/Queue/Update-cas-supplier-integration.sql [hunk 3]
```diff
@@ -0,0 +1,12 @@
+UPDATE
+    sims.queue_configurations
+SET
+    queue_configuration = '{
+        "cron": "0 19 * * 1-5",
```
Comment (dheepak-aot) [inline sources/packages/backend/apps/db-migrations/src/sql/Queue/Update-cas-supplier-integration.sql:5]: AC says noon. Either we can have the AC updated or change here. 

image

File: sources/packages/backend/apps/queue-consumers/src/processors/schedulers/cas-integration/_tests_/cas-supplier-integration.scheduler.e2e-spec.ts [hunk 4]
```diff
@@ -365,4 +365,68 @@ describe(describeProcessorRootTest(QueueNames.CASSupplierIntegration), () => {
...
     },
   );
+
+  it("Should throw an error for the first student and process the second one when the CAS API call failed for the first student but worked for the second one.", async () => {
+    // Arrange
+    // Create two mocks where the first one will throw an error
+    // and the second one is expected to work.
+    casServiceMock.getSupplierInfoFromCAS = jest
+      .fn()
+      .mockRejectedValueOnce("Unknown error")
```
Comment (dheepak-aot) [inline sources/packages/backend/apps/queue-consumers/src/processors/schedulers/cas-integration/_tests_/cas-supplier-integration.scheduler.e2e-spec.ts:375]: 👍

File: sources/packages/backend/libs/utilities/src/string-utils.ts [hunk 5]
```diff
@@ -140,5 +140,17 @@ export function convertToASCII(rawContent?: string): string | null {
...
+/**
+ * Converts input into an ASCII 7 bit buffer replacing special characters
+ * by equivalent ASCII characters when possible.
+ * @param rawContent raw content in string format.
+ * @returns a string of the input with extended ASCII characters (ISO-8859-1)
+ * converted to equivalent ASCII characters. If null or undefined is provided,
+ * null will be returned.
+ */
+export function convertToASCIIString(rawContent?: string): string | null {
+  return convertToASCII(rawContent)?.toString() ?? null;
```
Comment (dheepak-aot) [inline sources/packages/backend/libs/utilities/src/string-utils.ts:155]: Thanks for the change.

Comment (andrewsignori-aot) [reply sources/packages/backend/libs/utilities/src/string-utils.ts:155]: Thanks for raising the issue 😉

File: sources/packages/backend/apps/db-migrations/src/sql/Queue/Update-cas-supplier-integration.sql (hunk 3 above)
Comment (dheepak-aot) [reply sources/packages/backend/apps/db-migrations/src/sql/Queue/Update-cas-supplier-integration.sql:5]: My bad. I missed the PT part. As we discussed noon PST should be 8PM UTC.

Comment (andrewsignori-aot) [reply sources/packages/backend/apps/db-migrations/src/sql/Queue/Update-cas-supplier-integration.sql:5]: My bad too. All schedulers on wiki are using PST as a reference.

File: sources/packages/backend/apps/db-migrations/src/sql/Queue/Update-cas-supplier-integration.sql [hunk 6]
```diff
@@ -0,0 +1,12 @@
+UPDATE
+    sims.queue_configurations
+SET
+    queue_configuration = '{
+        "cron": "0 20 * * 1-5",
```
Comment (dheepak-aot) [inline sources/packages/backend/apps/db-migrations/src/sql/Queue/Update-cas-supplier-integration.sql:5]: Thanks. Please update the PR description as well.

**Review by dheepak-aot — APPROVED** [review]
Thanks for making the change. Looks good 👍

**Pull request merged**